Sources/PackageGraph/Pubgrub.swift

```diff
@@ -1233,41 +1258,62 @@ public final class PubgrubDependencyResolver {
     }
 
     /// Returns the best available version for a given term.
-    func getBestAvailableVersion(for term: Term) throws -> Version? {
+    func getBestAvailableVersion(for term: Term) throws -> BoundVersion? {
         assert(term.isPositive, "Expected term to be positive")
         let container = try getContainer(for: term.package)
 
         switch term.requirement {
         case .versionSet(let versionSet):
             let availableVersions = container.versions(filter: { versionSet.contains($0) } )
-            return availableVersions.first{ _ in true }
-        case .revision:
-            fatalError()
+            let version = availableVersions.first { _ in true }
+            return version.map(BoundVersion.version)
+        case .revision(let rev):
+            return .revision(rev)
         case .unversioned:
-            fatalError()
+            return .unversioned
         }
     }
 
     /// Returns the incompatibilities of a package at the given version.
     func incompatibilites(
         for package: PackageReference,
-        at version: Version
+        at version: BoundVersion
     ) throws -> [Incompatibility] {
         let container = try getContainer(for: package)
-        return try container.getDependencies(at: version).map { dep -> Incompatibility in
-            var terms: OrderedSet<Term> = []
-
-            // FIXME:
-            //
-            // If the selected version is the latest version, Pubgrub
-            // represents the term as having an unbounded upper range.
-            // We can't represent that here (currently), so we're
-            // pretending that it goes to the next nonexistent major
-            // version.
-            let nextMajor = Version(version.major + 1, 0, 0)
-            terms.append(Term(container.identifier, .versionSet(.range(version..<nextMajor))))
-            terms.append(Term(not: dep.identifier, dep.requirement))
-            return Incompatibility(terms, root: root!, cause: .dependency(package: container.identifier))
+
+        switch version {
+        case .version(let version):
+            return try container.getDependencies(at: version).map { dep -> Incompatibility in
+                var terms: OrderedSet<Term> = []
+
+                // FIXME:
+                //
+                // If the selected version is the latest version, Pubgrub
+                // represents the term as having an unbounded upper range.
+                // We can't represent that here (currently), so we're
+                // pretending that it goes to the next nonexistent major
+                // version.
+                let nextMajor = Version(version.major + 1, 0, 0)
+                terms.append(Term(container.identifier, .versionSet(.range(version..<nextMajor))))
+                terms.append(Term(not: dep.identifier, dep.requirement))
```

Comment:
Let's just `fatalError` for now if the requirement from a versioned dependency is not another versioned requirement.